src/runtime-rs/crates/resource/src/network/endpoint/veth_endpoint.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::io::{self, Error};
+
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use hypervisor::{device::NetworkConfig, Device, Hypervisor};
+
+use super::Endpoint;
+use crate::network::{utils, NetworkPair};
+
+#[derive(Debug)]
+pub struct VethEndpoint {
+    net_pair: NetworkPair,
+}
+
+impl VethEndpoint {
+    pub async fn new(
+        handle: &rtnetlink::Handle,
+        name: &str,
+        idx: u32,
+        model: &str,
+        queues: usize,
+    ) -> Result<Self> {
+        let net_pair = NetworkPair::new(handle, idx, name, model, queues)
+            .await
+            .context("new networkInterfacePair")?;
+        Ok(VethEndpoint { net_pair })
+    }
+}
+
+#[async_trait]
+impl Endpoint for VethEndpoint {
+    async fn name(&self) -> String {
+        self.net_pair.virt_iface.name.clone()
+    }
+
+    async fn hardware_addr(&self) -> String {
+        self.net_pair.tap.tap_iface.hard_addr.clone()
+    }
+
+    async fn attach(&self, h: &dyn Hypervisor) -> Result<()> {
+        let net_pair = &self.net_pair;
+        net_pair
+            .add_network_model()
+            .await
+            .context("add network model")?;
+        let iface = &net_pair.tap.tap_iface;
+        let guest_mac = utils::parse_mac(&iface.hard_addr).ok_or_else(|| {
+            Error::new(
+                io::ErrorKind::InvalidData,
+                format!("hard_addr {}", &iface.hard_addr),
+            )
+        })?;
+        let config = NetworkConfig {
+            id: net_pair.virt_iface.name.clone(),
+            host_dev_name: iface.name.clone(),
+            guest_mac: Some(guest_mac),
+        };
+        h.add_device(Device::Network(config))
+            .await
+            .context("Error add device")?;
+        Ok(())
+    }
+
+    async fn detach(&self, h: &dyn Hypervisor) -> Result<()> {
+        let net_pair = &self.net_pair;
+        net_pair
+            .del_network_model()
+            .await
+            .context("del network model")?;
+        let iface = net_pair.tap.tap_iface.clone();
```

Comment:
Lines 76-87 are almost the same as lines 52-63, so can create an fn to generate `NetworkConfig`.

Reply:
Thanks for your review, create a function `get_network_config`